File: apps/api-rs/src/document/quality.rs
use super::*;
pub(super) fn score_quality(parsed: &ParsedDocument) -> f64 {
    if parsed.blocks.is_empty() {
        return 0.2;
    }
    let char_count: usize = parsed.blocks.iter().map(|b| b.text.chars().count()).sum();
    let text_score = if char_count >= 500 {
        0.95
    } else if char_count >= 100 {
        0.8
    } else if char_count >= 20 {
        0.55
    } else {
        0.25
    };
    let invalid_chars = parsed
        .blocks
        .iter()
        .flat_map(|block| block.text.chars())
        .filter(|c| {
            *c == '\u{fffd}'
                || (c.is_control() && !matches!(*c, '\n' | '\r' | '\t'))
                || matches!(*c as u32, 0xe000..=0xf8ff | 0xf0000..=0xffffd | 0x100000..=0x10fffd)
        })
        .count();
    let legibility_score = if char_count == 0 {
        0.0
    } else {
        let invalid_ratio = invalid_chars as f64 / char_count as f64;
        if invalid_ratio >= 0.1 {
            0.15
        } else if invalid_ratio >= 0.02 {
            0.55
        } else {
            1.0
        }
    };
    let structure_score =
        if parsed.tables.is_empty() && !parsed.blocks.iter().any(|b| b.block_type == "heading") {
            0.7
        } else {
            0.95
        };
    let anchored_blocks = parsed
        .blocks
        .iter()
        .filter(|block| !block.anchor_ids.is_empty())
        .count();
    let anchor_score = anchored_blocks as f64 / parsed.blocks.len() as f64;
    let page_score = parsed.pages.map_or(1.0, |pages| {
        if pages <= 0 {
            return 0.0;
        }
        let covered = parsed
            .blocks
            .iter()
            .filter_map(|block| block.page_start.or(block.slide_index))
            .collect::<std::collections::BTreeSet<_>>()
            .len();
        (covered as f64 / pages as f64).clamp(0.0, 1.0)
    });
    let warning_penalty = quality_warning_penalty(&parsed.warnings);
    (0.3 * text_score
        + 0.2 * legibility_score
        + 0.15 * structure_score
        + 0.2 * anchor_score
        + 0.15 * page_score
        - warning_penalty)
        .clamp(0.0, 1.0)
}

fn quality_warning_penalty(warnings: &[String]) -> f64 {
    let mut penalty = warnings.len() as f64 * 0.02;
    for warning in warnings {
        if warning == "scanned_pdf_no_text_layer" {
            penalty += 0.4;
        } else if let Some(ratio) = warning.strip_prefix("pdf_partial_text_layer:") {
            let missing_ratio = ratio
                .split_once('/')
                .and_then(|(missing, total)| {
                    Some(missing.parse::<f64>().ok()? / total.parse::<f64>().ok()?)
                })
                .unwrap_or(0.0);
            penalty += if missing_ratio >= 0.15 { 0.15 } else { 0.05 };
        }
    }
    penalty.min(0.5)
}

File: apps/api-bun/src/document/quality.ts
// 移植自 apps/api-rs/src/document/quality.rs —— 行为对齐 Rust 原版，错误信息保持一致

// 文档质量评分：文本量、可读性、结构、锚点覆盖、页覆盖率与警告惩罚

import { charCount } from './text_utils.ts';
import type { ParsedDocument } from './types.ts';

export function scoreQuality(parsed: ParsedDocument): number {
  if (parsed.blocks.length === 0) return 0.2;
  const charTotal = parsed.blocks.reduce((sum, block) => sum + charCount(block.text), 0);
  const textScore = charTotal >= 500 ? 0.95
    : charTotal >= 100 ? 0.8
      : charTotal >= 20 ? 0.55
        : 0.25;
  let invalidChars = 0;
  for (const block of parsed.blocks) {
    for (const ch of block.text) {
      if (isInvalidChar(ch)) invalidChars += 1;
    }
  }
  let legibilityScore: number;
  if (charTotal === 0) {
    legibilityScore = 0.0;
  } else {
    const invalidRatio = invalidChars / charTotal;
    legibilityScore = invalidRatio >= 0.1 ? 0.15
      : invalidRatio >= 0.02 ? 0.55
        : 1.0;
  }
  const structureScore = parsed.tables.length === 0
    && !parsed.blocks.some((block) => block.block_type === 'heading')
    ? 0.7
    : 0.95;
  const anchoredBlocks = parsed.blocks.filter((block) => block.anchor_ids.length > 0).length;
  const anchorScore = anchoredBlocks / parsed.blocks.length;
  let pageScore = 1.0;
  if (parsed.pages !== null) {
    if (parsed.pages <= 0) {
      pageScore = 0.0;
    } else {
      const covered = new Set<number>();
      for (const block of parsed.blocks) {
        const marker = block.page_start ?? block.slide_index;
        if (marker !== null) covered.add(marker);
      }
      pageScore = Math.min(1.0, Math.max(0.0, covered.size / parsed.pages));
    }
  }
  const warningPenalty = qualityWarningPenalty(parsed.warnings);
  const score = 0.3 * textScore
    + 0.2 * legibilityScore
    + 0.15 * structureScore
    + 0.2 * anchorScore
    + 0.15 * pageScore
    - warningPenalty;
  return Math.min(1.0, Math.max(0.0, score));
}

function isInvalidChar(ch: string): boolean {
  const codePoint = ch.codePointAt(0)!;
  if (codePoint === 0xfffd) return true;
  // Rust char::is_control：Cc 类别
  if ((codePoint >= 0x00 && codePoint <= 0x1f) || (codePoint >= 0x7f && codePoint <= 0x9f)) {
    return !['\n', '\r', '\t'].includes(ch);
  }
  if (codePoint >= 0xe000 && codePoint <= 0xf8ff) return true;
  if (codePoint >= 0xf0000 && codePoint <= 0xffffd) return true;
  if (codePoint >= 0x100000 && codePoint <= 0x10fffd) return true;
  return false;
}

function qualityWarningPenalty(warnings: string[]): number {
  let penalty = warnings.length * 0.02;
  for (const warning of warnings) {
    if (warning === 'scanned_pdf_no_text_layer') {
      penalty += 0.4;
    } else if (warning.startsWith('pdf_partial_text_layer:')) {
      const ratio = warning.slice('pdf_partial_text_layer:'.length);
      const slash = ratio.indexOf('/');
      let missingRatio = 0.0;
      if (slash >= 0) {
        const missing = Number.parseFloat(ratio.slice(0, slash));
        const total = Number.parseFloat(ratio.slice(slash + 1));
        if (Number.isFinite(missing) && Number.isFinite(total) && total !== 0) {
          missingRatio = missing / total;
        }
      }
      penalty += missingRatio >= 0.15 ? 0.15 : 0.05;
    }
  }
  return Math.min(penalty, 0.5);
}
